kdc/krb5tgs.c

```diff
@@ -1379,78 +1394,7 @@ tgs_build_reply(astgs_request_t priv,
     if (b->kdc_options.canonicalize)
 	flags |= HDB_F_CANON;
 
-    if(b->kdc_options.enc_tkt_in_skey){
-	Ticket *t;
-	hdb_entry_ex *uu;
-	krb5_principal p;
-	Key *uukey;
-	krb5uint32 second_kvno = 0;
-	krb5uint32 *kvno_ptr = NULL;
-
-	if(b->additional_tickets == NULL ||
-	   b->additional_tickets->len == 0){
-	    ret = KRB5KDC_ERR_BADOPTION; /* ? */
-	    kdc_log(context, config, 4,
-		    "No second ticket present in user-to-user request");
-            _kdc_audit_addreason((kdc_request_t)priv,
-                                 "No second ticket present in user-to-user request");
-	    goto out;
-	}
-	t = &b->additional_tickets->val[0];
-	if(!get_krbtgt_realm(&t->sname)){
-	    kdc_log(context, config, 4,
-		    "Additional ticket is not a ticket-granting ticket");
-            _kdc_audit_addreason((kdc_request_t)priv,
-                                 "Additional ticket is not a ticket-granting ticket");
-	    ret = KRB5KDC_ERR_POLICY;
-	    goto out;
-	}
-	_krb5_principalname2krb5_principal(context, &p, t->sname, t->realm);
-	ret = krb5_unparse_name(context, p, &tpn);
-	if (ret)
-		goto out;
-	if(t->enc_part.kvno){
-	    second_kvno = *t->enc_part.kvno;
-	    kvno_ptr = &second_kvno;
-	}
-	ret = _kdc_db_fetch(context, config, p,
-			    HDB_F_GET_KRBTGT, kvno_ptr,
-			    NULL, &uu);
-	krb5_free_principal(context, p);
-	if(ret){
-	    if (ret == HDB_ERR_NOENTRY)
-		ret = KRB5KDC_ERR_S_PRINCIPAL_UNKNOWN;
-            _kdc_audit_addreason((kdc_request_t)priv,
-                                 "User-to-user service principal (TGS) unknown");
-	    goto out;
-	}
-	ret = hdb_enctype2key(context, &uu->entry, NULL,
-			      t->enc_part.etype, &uukey);
-	if(ret){
-	    _kdc_free_ent(context, uu);
-	    ret = KRB5KDC_ERR_ETYPE_NOSUPP; /* XXX */
-            _kdc_audit_addreason((kdc_request_t)priv,
-                                 "User-to-user enctype not supported");
-	    goto out;
-	}
-	ret = krb5_decrypt_ticket(context, t, &uukey->key, &adtkt, 0);
-	_kdc_free_ent(context, uu);
-	if(ret) {
-            _kdc_audit_addreason((kdc_request_t)priv,
-                                 "User-to-user TGT decrypt failure");
-	    goto out;
-        }
-
-	ret = verify_flags(context, config, &adtkt, tpn);
-	if (ret) {
-            _kdc_audit_addreason((kdc_request_t)priv,
-                                 "User-to-user TGT expired or invalid");
-	    goto out;
-        }
-
-	s = &adtkt.cname;
-	r = adtkt.crealm;
-    } else if (s == NULL) {
+    if (s == NULL) {
```

Comment:
Isn't an empty server allowed for U2U?

Reply:
I agree the draft at https://datatracker.ietf.org/doc/html/draft-ietf-cat-user2user-02 says it is optional however on Windows  our tests show it gives KDC_ERR_GENERIC  or  KDC_ERR_S_PRINCIPAL_UNKNOWN (probably GENERIC, MS is not good with error codes). https://gitlab.com/samba-team/samba/-/commit/26480ba2aa9834a24f1ea11ae3f8e2d7ed0ccfd8

@josephsutton1 may be able to shed some light on the tests as they are his tests.

Reply:
Yes, it gives a GENERIC error on Windows. The commit 'Check name in request against name in user-to-user TGT' states that the reason to require the server name is so it can be checked against the client name in the ticket.

Reply:
OK, it looks like MIT also requires the server principal to be non-NULL in all cases.

Reply:
After discussing with @nicowilliams, yes, we do need to check that the additional ticket clients matches the server name in all cases.